README.md

# MediaPlayer

Simple MediaPlayer App that able to Scan &amp; Play local playable Media. Using Exoplayer with Dagger-hilt &amp; ViewModels

Kind of ready, pretty much usable

> - If you want to clone this simply remove the BuildConfig check in `MPApplication`
> - ~~Audio should be inside the `Music` or Folder Generated by the OS~~ Android 29 / Q only
> - Audio Images are fetched from the Album Art and not the one embedded in the files. `Might support this in the future with ID3v2`
> - Android Nougat 7.0 / API 24 and above

Features :
- Foreground Service on Background & Lock (tested 29 & 30 only)
- Playlist From Album & Artist row in `Home`
- MediaStore Audio Files Scan
- Playing Fragment with Palette Api & Glide transformations
- Folder Path
- Viewpager display currently playing song along with Album Images

Working On:
- Artist & Album Fragment
- Better Transport Control
- Better UI / UX


Plans :
- History, Favorite, Playlist, Album Image Changer with Room & DataStore. `> ID3v2 If Possible`
- Video Player support with                                               `> PiP, Brightness by Scroll, Volume by Scroll, Seekbar Preview`
- Media Streaming with Firebase.                                          `> Got Firebase Storage Limit so might be just Downloadable Instead`
- Google & Others Login handled by Firebase.                              `> Might add Local Account but most likely not `
- Custom Theming

Credits :
- Building feature-rich media apps with ExoPlayer (Google I/O '18)

- Uamp
- Spotify Clone
- Music-Player-GO
- RetroMusicPlayer
- Flaticon

| Light Mode |
| -------------- |
|<img src="https://user-images.githubusercontent.com/94031495/153752616-3becefe9-fa6e-4d0e-885a-984885bd1394.png" width="250"> <img src="https://user-images.githubusercontent.com/94031495/153752761-3f698278-d2dd-4929-ae24-0931ab5b04ac.png" width="250"> <img src="https://user-images.githubusercontent.com/94031495/153752862-cde61153-ad2a-4d38-a9fc-5a302cbd5ac6.png" width="250"> |

| Dark Mode |
| -------------- |
| <img src="https://user-images.githubusercontent.com/94031495/153753081-eed0709b-fe53-4a41-83fa-c3a91d5add06.png" width="250"> <img src="https://user-images.githubusercontent.com/94031495/153753082-9fd3dfe0-dbc2-4570-a767-ea1bbe4640e2.png" width="250"> <img src="https://user-images.githubusercontent.com/94031495/153753083-2ee6c2ba-6ab2-4809-90db-7a67b3ba3920.png" width="250"> |


Thank you For Visiting ^_^\
· I'm Usually on https://discord.gg/programming if you have question


Changelog.md
#2022/02/13
- Fix Bug where MainActivity Swipe before songList is Updated

#2022/02/12
- Fix Bug when Item of playlist is deleted and will try to recover

#2022/02/11
- PlayingFragment ViewPager Loop
- seekTo Percentage instead of Duration
- Seekbar Object Animator
- Fix Notification Channel Bug on PlayerNotification

#2022/02/08
- PlayingFragment Palette API background
- reload Children on Device Song List changes

#2022/02/07
-  PlayingFragment that shows currently Playing Song
-  PlayerNotification & PlayingFragment Repeat State
-  PlayingFragment seekTo with Seekbar
-  PlayingFragment Skip Next / Prev button
-  PlayingFragment observe Duration & Progress
-  Reduce Transition Lag

# ~~~~~~~~~~~~

# 2022/02/03
- Added CustomActionReceiver on PlayerNotificationManager
- Added Notification Repeat button that Toggle the exoPlayer Repeat state between OFF, ONCE, ALL
- Notification Repeat button will change depending on the Player Repeat state
- Notification Repeat button will now be placed at index 0 instead the default one defined `usually at [3]`
- now can set Playlist with SearchView by taking the shown songList

# 2022/02/02
- App will now send Toast if needed and not simply crash when exception happen
- Added delay between ServiceConnector request, will cancel if its in between Cooldown
- Fixed Navigator inconsistency when sending command too fast to Connector
- Fixed another inconsistency on MediaStore Query & ViewModels Livedata should only be observed
- Fixed Folder not distinct'd

# 2022/02/01
- Added temporary playlist by Artist & Album from Home Page
- Added Notification Controller!!
- fixed another `onLoadChildren Exception` when changing playlist
- fixed ViewPager not Gliding when changing playlist
- fixed playlist cleared when Notification Cancelled
- fixed suggestion empty when Service Stopped
- fixed yet another liveData NPE, liveData reference will be moved to MainActivity observer instead
- fixed musicQueueNavigator Exception
- added command method to interact with MusicService from UI-VM/Connector-Service

# 2022/01/31
- Fixed ViewPager not Gliding when Activity is Recreated
- Moved Media Query to Repo
- Better Playback Controller
- Fix onLoadChildren Exception
- Fix NPE everywhere :)

# 2022/01/30
- added MusicSource class for converting fetched song metadata to MediaMetadataCompat
- setting up NotificationManager class for handling Notification by MusicService Class
- added NotificationListener class to Listen for NotificationManager changes / callbacks
- added MusicService Playback Preparer & Listener
- added MusicServiceConnector as connector between Activity/ViewModel with MusicService

# 2022/01/29
- yet Another fix to Player Inconsistency `will migrate the player to a Service`
- yet Another fix to problem when song is deleted in Storage `will migrate using Room & Datastore`
- the simple controller will now check for player state even after Activity Recreation `temporary before Service Implemented`
- added better transition & ripple onclick effect
- a little cleanup on MainActivity & ViewModels

# 2022/01/28
- Fixed Activity FinishCallback Leak on Android Q
- Fixed Recyclerview Padding when hiding bnv
- Added Simple Player Listener for playback state & seekbar `Temporary`
- Ability to Play from Home & Folder Fragment
- Fixed Play / Pause Inconsistency when Device Theme changed & Activity Recreation

# 2022/01/27
- Fixed Leak on Adapter & Coordinator Layout
- Fixed Shuffled Song on Suggestion Row is Persisting after Song Deletion on Storage  
- Change Fragment Transition to Motion Layout from Material Design
- Tested and expected the App to be working on Android Nougat 7.0 and above
- Change build gradle to 7.1.0 as per Android Studio Bumblebee 2021.1.1 Stable Update

app/src/main/java/com/example/mediaplayer/util/Helper.kt
package com.example.mediaplayer.util

import android.content.Context
import android.os.Build
import com.example.mediaplayer.util.Constants.FOREGROUND_SERVICE
import com.example.mediaplayer.util.Constants.PERMISSION_FOREGROUND_SERVICE_REQUEST_CODE
import com.example.mediaplayer.util.Constants.PERMISSION_WRITE_EXT_REQUEST_CODE
import com.example.mediaplayer.util.Constants.WRITE_STORAGE
import com.vmadalin.easypermissions.EasyPermissions
import timber.log.Timber

object VersionHelper {
    fun isNougat() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N     // Nougat       7.0     24
    fun isNMR() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.N_MR1    // Nougat       7.1     25
    fun isOreo() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O       // Oreo         8.0     26
    fun isOMR() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1    // Oreo         8.1     27
    fun isPie() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P        // Pie          9.0     28
    fun isQ() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q          // Q            10      29
    fun isR() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.R          // Red Velvet   11      30
    fun isSnowCone() = Build.VERSION.SDK_INT >= Build.VERSION_CODES.S   // Snow Cone    12      31
}

object Version {
    fun isNougat() = Build.VERSION.SDK_INT == Build.VERSION_CODES.N     // Nougat       7.0     24
    fun isNMR() = Build.VERSION.SDK_INT == Build.VERSION_CODES.N_MR1    // Nougat       7.1     25
    fun isOreo() = Build.VERSION.SDK_INT == Build.VERSION_CODES.O       // Oreo         8.0     26
    fun isOMR() = Build.VERSION.SDK_INT == Build.VERSION_CODES.O_MR1    // Oreo         8.1     27
    fun isPie() = Build.VERSION.SDK_INT == Build.VERSION_CODES.P        // Pie          9.0     28
    fun isQ() = Build.VERSION.SDK_INT == Build.VERSION_CODES.Q          // Q            10      29
    fun isR() = Build.VERSION.SDK_INT == Build.VERSION_CODES.R          // Red Velvet   11      30
    fun isSnowCone() = Build.VERSION.SDK_INT == Build.VERSION_CODES.S   // Snow Cone    12      31
}

object PermsHelper {
    fun checkStoragePermission(context: Context) =
        hasPermission(Perms(WRITE_STORAGE, PERMISSION_WRITE_EXT_REQUEST_CODE), context)
    fun checkForegroundServicePermission(context: Context) =
        if (VersionHelper.isPie()) hasPermission(
            Perms(FOREGROUND_SERVICE, PERMISSION_FOREGROUND_SERVICE_REQUEST_CODE), context
        ) else true

    private fun hasPermission(perms: Perms, context: Context): Boolean {
        return try {
            Timber.d("hasPermission ?: $perms")
            EasyPermissions.hasPermissions(context, perms.permission)
        } catch (e: Exception) {
            Timber.wtf(RuntimeException("$e $perms"))
            false
        }
    }
}

data class Perms(
    val permission: String,
    val requestId: Int = 130,
    val msg: String? = null
)